src/extension/token/minting/BurnAndMintERC721.sol

```diff
@@ -0,0 +1,286 @@
+    {
+        SaleConfig memory saleConfig = _burnAndMintStorage().saleConfig[_token];
+        return (saleConfig.primarySaleRecipient, saleConfig.platformFeeRecipient, saleConfig.platformFeeBps);
+    }
+
+    /// @notice Sets the sale configuration for a token.
+    function setSaleConfig(address _primarySaleRecipient, address _platformFeeRecipient, uint16 _platformFeeBps)
+        external
+    {
+        address token = msg.sender;
+        _burnAndMintStorage().saleConfig[token] =
+            SaleConfig(_primarySaleRecipient, _platformFeeRecipient, _platformFeeBps);
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                            INTERNAL FUNCTIONS
+    //////////////////////////////////////////////////////////////*/
+
+    /// @notice Mint new tokens after burning required tokens from origin contract.
+    function _burnAndMint(address _tokenOwner, uint256 _quantity, BurnAndMintParams memory _params) internal {
+        BurnAndMintConditions memory _conditions = _burnAndMintStorage().conditions[msg.sender];
+
+        // Verify and burn tokens on origin contract
+        _verifyBurnAndMint(_tokenOwner, _params.burnTokenId, _quantity, _conditions);
+        _burnTokensOnOrigin(_tokenOwner, _params.burnTokenId, _quantity, _conditions);
+
+        // Collect price
+        if (_conditions.currency != address(0)) {
+            _distributeMintPrice(_tokenOwner, _conditions.currency, _quantity * _conditions.mintPriceForNewToken);
+        }
+
+        // emit event
+        emit TokensBurnedAndClaimed(_conditions.originContractAddress, _tokenOwner, _params.burnTokenId, _quantity);
+    }
+
+    function _verifyBurnAndMint(
+        address _tokenOwner,
+        uint256 _tokenId,
+        uint256 _quantity,
+        BurnAndMintConditions memory _conditions
+    ) public view virtual {
+        if (_conditions.tokenType == TokenType.ERC721) {
+            if (_quantity != 1) {
+                revert BurnAndMintInvalidQuantity();
+            }
+            if (IERC721(_conditions.originContractAddress).ownerOf(_tokenId) != _tokenOwner) {
+                revert BurnAndMintNotOwner();
+            }
+        } else if (_conditions.tokenType == TokenType.ERC1155) {
+            uint256 _eligible1155TokenId = _conditions.tokenId;
+
+            if (_tokenId != _eligible1155TokenId) {
+                revert BurnAndMintInvalidTokenId();
+            }
+
+            if (IERC1155(_conditions.originContractAddress).balanceOf(_tokenOwner, _tokenId) < _quantity) {
+                revert BurnAndMintNotEnoughBalance();
+            }
+        }
+    }
+
+    function _burnTokensOnOrigin(
+        address _tokenOwner,
+        uint256 _tokenId,
+        uint256 _quantity,
+        BurnAndMintConditions memory _conditions
+    ) internal virtual {
+        if (_conditions.tokenType == TokenType.ERC721) {
+            IERC721Burnable(_conditions.originContractAddress).burn(_tokenId);
```

Comment:
The ERC721Core and ERC1155Core contracts have a different function signature for burn, which means they are not supported as origin contract on this extension.

We should use some method e.g. try-catch to support contracts with and without bytes data in their function signature